key-value attrs should be arrayref of arrayrefs

bpj opened the pull request:
*   Pandoc expects `$attr->[2]` to be an arrayref of arrayrefs, each with a key-value pair,
    whereas Pandoc::Elements::attributes() used to create a flat arrayref.
    The function and tests have been updated to produce the format exported and expected by Pandoc.

*   Added convenience methods `keyval_pairs`, `keyval_list` and `attr_list` to
    simplify access to attributes and creation of Hash::MultiValue objects with the
    corrected LoL key-value pairs format.

*   Also added a line to silence some "only used once" syntax check warnings.

nichtich commented:
Perl 2.22 features or why do this fail?
https://travis-ci.org/nichtich/Pandoc-Elements/builds/114542929

nichtich commented:
I'd also rename the method names but finding good names is hard ;-)

bpj commented:
Apparently they have pandoc installed on one of their servers. That's why one of the tests passed. I added a conditional test checking that pandoc accepts the attributes, but forgot to check for pandoc itself! 
I added a check that `$ENV{PANDOC_VERSION} exists. Maybe I should remove that test entirely, but I figure that nobody who doesn't have pandoc installed will set that variable, but clearly that test should be an xt test! I got a bit carried away.

bpj commented:
Cleared up now. Sorry for the confusion.

BTW there is another problem with attributes: pandoc won't accept values which are numbers as opposed to strings.  I wrote a fix but it makes the attributes an object so that they can have a TO_JSON method which enforces stringification.  Are you interested anyway?

nichtich commented:
Thanks! I'd also prefer making the attributes a blessed object but its internal structure (which you fixed) should stay the same. So `$element->attr` returns an object with additional methods. Accessor methods should be compatible to some degree with Hash::MultiValue, e.g.

* `$element->attr->get($key)` returns a single value for the given $key. If there are multiple values, the last one.
* `$element->attr->get_all($key)` returns a list of values for the given $key
* `$element->attr->keys` returns a list of all keys, including duplicates

Other possible methods (nice to have):

* `$element->attr->pairs` (to be passed to Hash::MultiValue->new)
* `$element->attr->set_all` (to set from Hash::MultiValue->flatten)
* `$element->attr->set($key [, $value ... ])` changes/adds/removes value(s) for given key, preserving order
* `$element->attr->unique_keys`

bpj commented:
In [my branch][] I have a [modified attributes function][] which accepts a hashref *or* a flat arrayref of pairs as argument, with either strings or arrayrefs as values. Values of `classes` keys are automatically merged into a single arrayref, while arrayref values of other keys (except `id` of course) are unpacked as pairs (the latter also for hashref input, so that it will do the right thing with the output of Hash_MultiValue::as_hashref_mixed).

Some of the methods you suggest do the same thing as the methods I added to \*AttributesRole. Probably they should be methods of the \*Attributes object instead (or as well).

If we are to keep the pandoc-like internal structure of the \*Attributes object those modifying methods preserving order are going to be tricky! They would be a little easier to do if we can use List::MoreUtils though. I like to believe that anybody who uses Perl seriously has L::MU installed, and there already is JSON as a non-core dependency.

What would set() do if there are more new than old values for an existing key? Replace the existing ones and add the 'superfluous' ones at the end?

[my branch]: https://github.com/bpj/Pandoc-Elements/tree/bpj-attribute-stringification
[modified attributes function]: https://github.com/bpj/Pandoc-Elements/blob/bpj-attribute-stringification/lib/Pandoc/Elements.pm#L139

nichtich commented:
Sorry, I merged by accident. I prefer to cherry-pick. The issue with number values can occur everywhere, here is a test:

~~~perl
my $ast = element ( Header => 6, attributes { foo => 6 }, Str 6 );
is_deeply [ $ast->to_json =~ /(.6.)/g ], [ '"6"', '"6"', '"6"',],  'stringify numbers';
~~~

I'll have a closer look Monday.

bpj commented:
Yes, the problem is that in the level number of a Header pandoc actually wants a number so it must be coerced in TO_JSON with `0+$level`. The same is true of the column widths in a table and the start number of an ordered list, and maybe a few other places. In the content of a MetaBool it wants a true Boolean, so that must be coerced to a JSON::PP::Boolean object with something like `$value ? JSON::true() : JSON::false()`. Everywhere else pandoc expects scalars to be strings, so they must be coerced with `"$value"` or `"" . $value`. The right place to do this is clearly in the TO_JSON method. I'll look into this during the week. I have no clients ATM so I have time. I made a document once which exemplifies all pandoc elements. I'll update it, convert it to JSON with pandoc, prettyprint it and note all places where pandoc doesn't expect scalar values to be strings and write/update bespoke TO_JSON methods as needed. I can't make up my mind whether Perl is too lax with scalar types or if pandoc/Haskell is too strict. I wish Perl had proper coercion functions though.

I have a question: do you think TO_JSON should/may coerce the data in its invocant object and then make a shallow copy or should it copy everything? In the past I've used Data::Rmap to walk the whole structure coercing all scalars to strings and then recoerced some to numbers as appropriate. I can write a helper routine which coerces all scalars in a data structure to strings or do it in Pandoc::Document::Element::TO_JSON and then have bespoke TO_JSON methods first call SUPER::TO_JSON and then recoerce to numbers as needed. That's easier than having bespoke TO_JSON methods for all element classes.

bpj commented:
Something like this

    my $clone_data = sub {  # has file scope so visible in all inner packages
        my($data, $clone_data) = @_;
        if ( blessed $data ) {
            return $data if $data->can('TO_JSON'); # JSON.pm will convert
            return "$data";  # might have overloaded stringification!
        }
        elsif ( 'ARRAY' eq ref $data ) {
            return [  
                map { 
                    ref($_) ? $clone_data->($_, $clone_data) : "$_";
                } @$data 
            ];
        }
        elsif ( 'HASH' eq ref $data ) {
            my %ret = %$data;
            while ( my( $I, $v ) = each %ret ) {
                $ret{$k} = ref($v) ? $clone_data->($v, $clone_data) : "$v";
            }
            return \%ret;
        }
        else { return "$data" }
    };
    
    sub *::Element::TO_JSON {
        return $clone_data->( { %{ $_[0] } }, $clone_data );
    }
    
    sub *::Table::TO_JSON {
        my $data = $clone_data->( { %{ $_[0] } }, $clone_data );
        my $widths = $data->{c}[2];
        @$widths = map { 0+$_ } @$widths;
        return $data;
    }
  
    sub *::MetaBool::TO_JSON {
        my $data = { %{ $_[0] } };
        $data->{c} = $data->{c} ? JSON::true() : JSON::false();
        return $data;
    }
  
    sub *::Header::TO_JSON {
        my $data = $clone_data->( { %{ $_[0] } }, $clone_data );
        $data->{c}[0] = 0+ $data->{c}[0];
        return $data;
    }

nichtich commented:
Looks complicated but given that JSON data type varies among elements and fields, we must implement TO_JSON for each case. Modification of types in-place should be better for large documents and to keep the internal data structure same as Pandoc JSON structure but that my be more difficult to implement.

I won't have time to review before easter. To answer questions on the other issues:

> Do you think it would be a good idea to check for duplicate class values? Or duplicate values generally?

No check. Pandoc supports duplicate class values as well:

    echo '# {.x .x}' | pandoc -t json
    [{"unMeta":{}},[{"t":"Header","c":[1,["section",["x","x"],[]],[]]}]]

> What would set() do if there are more new than old values for an existing key? Replace the existing ones and add the 'superfluous' ones at the end?

Yes I think that's what Hash::MultiValue also does.

nichtich commented:
This discussion thread mixes too many issues so I'll cherry pick one by one.

nichtich commented:
In 0.19 I added the `keyvals` method with Hash::MultiValue for easier access to attributes. Write access has not been implemented (#30).
